docs/spec/store/interblock-cache.md

```diff
@@ -0,0 +1,294 @@
+
+## System model and properties
+
+### Assumptions
+
+This specification assumes that there exists a cache implementation accessible to the inter-block cache feature.
+
+> The implementation uses adaptive replacement cache (ARC), an enhancement over the standard last-recently-used (LRU) cache in that tracks both frequency and recency of use.
+
+The inter-block cache requires that the cache implementation to provide methods to create a cache, add a key/value pair, remove a key/value pair and retrieve the value associated to a key. In this specification, we assume that a `Cache` feature offers this functionality through the following methods:
+
+* `NewCache(size: int)` creates a new cache with `size` capacity and returns it.
+* `Get(key: string)` attempts to retrieve a key/value pair from `Cache.` It returns `[value: []byte, success: bool]`. If `Cache` contains the key, it `value` contains the associated value and `success=true`. Otherwise, `success=false` and `value` should be ignored.
+* `Add(key: string, value: []byte)` inserts a key/value pair into the `Cache`.
+* `Remove(key: string)` removes the key/value pair identified by `key` from `Cache`.
+
+The specification also assumes that `CommitKVStore` offers the following API:
+
+* `Get(key: string)` attempts to retrieve a key/value pair from `CommitKVStore`.
+* `Set(key, string, value: []byte)` inserts a key/value pair into the `CommitKVStore`.
+* `Delete(key: string)` removes the key/value pair identified by `key` from `CommitKVStore`.
+
+> Ideally, both `Cache` and `CommitKVStore` should be specified in a different document and referenced here.
+
+### Properties
+
+#### Thread safety
+
+Accessing the `cache manager` or a `KVCache` is not thread-safe: no method is guarded with a lock.
+Note that this is true even if the cache implementation is thread-safe.
+
+> For instance, assume that two `Set` operations are executed concurrently on the same key, each writing a different value. After both are executed, the cache and the underlying store may be inconsistent, each storing a different value under the same key.
+
+#### Crash recovery
+
+The inter-block cache transparently delegates `Commit()` to its aggregate `CommitKVStore`. If the 
+aggregate `CommitKVStore` supports atomic writes and use them to guarantee that the store is always in a consistent state in disk, the inter-block cache can be transparently moved to a consistent state when a failure occurs.
+
+> Note that this is the case for `IAVLStore`, the preferred `CommitKVStore`. On commit, it calls `SaveVersion()` on the underlying `MutableTree`. `SaveVersion` writes to disk are atomic via batching. This means that only consistent versions of the store (the tree) are written to the disk. Thus, in case of a failure during a `SaveVersion` call, on recovery from disk, the version of the store will be consistent.
+
+#### Iteration
+
+Iteration is not supported.
```

Comment:
In the implementation, since `CommitKVStoreCache` embeds a `CommitKVStore`, it does expose the `CommitKVStore` iteration API.

Since this is just a system model, I guess it's sound to underspecify here, but probably surprising to the reader?

Reply:
I've added a sentence. Please check.

Reply:
Looks good, thx!